server/src/main/java/io/crate/expression/reference/doc/lucene/ColumnFieldVisitor.java

```diff
@@ -0,0 +1,216 @@
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.jetbrains.annotations.NotNull;
+
+import io.crate.common.collections.Maps;
+import io.crate.execution.dml.ArrayIndexer;
+import io.crate.metadata.ColumnIdent;
+import io.crate.metadata.Reference;
+import io.crate.metadata.doc.DocTableInfo;
+import io.crate.metadata.doc.SysColumns;
+import io.crate.types.ArrayType;
+import io.crate.types.DataType;
+import io.crate.types.GeoShapeType;
+import io.crate.types.ObjectType;
+
+/**
+ * Loads column values from stored fields and presents them as a java Map
+ */
+public class ColumnFieldVisitor extends StoredFieldVisitor {
+
+    private final Map<String, Field> fields = new HashMap<>();
+    private final Set<ColumnIdent> droppedColumns;
+    private final SourceParser storedSourceParser;
+
+    /**
+     * Creates a new ColumnFieldVisitor for the given table
+     */
+    public ColumnFieldVisitor(DocTableInfo table) {
+        this.droppedColumns
+            = table.droppedColumns().stream().map(Reference::column).collect(Collectors.toUnmodifiableSet());
+        this.storedSourceParser = new SourceParser(table);
+    }
+
+    private interface Field extends Comparable<Field> {
+
+        Object sanitize(Object v);
+
+        ColumnIdent column();
+
+        default int compareTo(@NotNull ColumnFieldVisitor.Field o) {
+            return column().compareTo(o.column());
+        }
+    }
+
+    private record ValueField(DataType<?> dataType, ColumnIdent column) implements Field {
+        @Override
+        public Object sanitize(Object v) {
+            return dataType.sanitizeValue(v);
+        }
+    }
+
+    private record ArrayOfObjectField(DataType<?> dataType, ObjectType objectType, ColumnIdent column, SourceParser sourceParser) implements Field {
+        @Override
+        public Object sanitize(Object v) {
+            try {
+                var map = sourceParser.parse(new BytesArray((byte[]) v), Map.of(column.leafName(), objectType.innerTypes()), false);
+                if (map.isEmpty()) {
+                    return List.of();
+                }
+                return map.values().iterator().next();
+            } catch (NotXContentException e) {
+                // may be an array of nulls inserted before the field was upcast to an array of objects
+                try (StreamInput in = new ByteBufferStreamInput(ByteBuffer.wrap((byte[]) v))) {
+                    return in.readList(StreamInput::readGenericValue);
+                } catch (IOException io) {
+                    throw new UncheckedIOException(io);
+                }
+            }
+        }
+    }
+
+    private record GeoShapeField(DataType<?> dataType, ColumnIdent column, SourceParser sourceParser) implements Field {
+        @Override
+        public Object sanitize(Object v) {
+            var map = sourceParser.parse(new BytesArray((byte[]) v), true);
+            return map.values().iterator().next();
+        }
+    }
```

Comment:
A bit related to my comment regarding having a read/decode in DataType#StorageSupport. Is there a way we could avoid the type specific logic here?

Object/Array special handling is to some degree okay because they're containers, but geo-shape should mostly work like any other value. (We might also change the representation in the future away from a Map, to have concrete properties)